drafts/AENS.md

```diff
@@ -0,0 +1,414 @@
+can be understood as the `mywallet` namespace in the `aet` namespace.
+
+A label MUST be longer than three characters.
+
+Labels SHOULD be at most 63 characters long and the full path
+of all names and namespaces SHOULD be at most 253 characters
+long, in order to enable interoperability with DNS.
+
+
+### Registrars
+
+A registrar controls who is allowed to claim names in their namespace
+and under what circumstances they are allowed to do so.
+
+The only available registrars for this AENS version will be hard-coded
+ones, which own the `.`, `.aet` and `.test` namespaces.
+
+The `.` namespace registrar is restricted and MUST NOT allow anyone to
+claim any names in its namespace.
+
+The `.aet` namespace registrar allows users to claim any name that is
+valid according to the validation rules for labels and names if they
+are the first to claim the name and pay the appropriate fee.
+
+The `.test` namespace registrar can be, as the name suggests, used for
+test purposes
+
+
+### Commitment Scheme
+
+Claiming a name requires a commitment scheme, which enables the claimant
+to commit to a name and after the commitment has been accepted into the 
+chain, reveal the name to finish the process.
+
+A commitment should be binding, i.e. the claimant cannot change
+the value they commited to withouth changing the actual commitment
+and hiding, so that a malicious miner learns nothing about the value
+the claimant has commited until they chose to reveal that value. This
+prevents malicious miners from front running, i.e. upon seeing a 
+claim transaction, including their own claim request for the same
+name instead of the original claimant's one.
+
+
+### Hashing
+
+We are going to adopt the `NameHash` as defined by the ENS in 
+[EIP-137](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-137.md#namehash-algorithm)
+using SHA256 instead.
+
+```
+def namehash(name):
+  if name == '':
+    return '\0' * 32
+  else:
+    label, _, remainder = name.partition('.')
+    return sha256(namehash(remainder) + sha256(label))
+```
+
+Names are generally only referred to in hashed form.
+
+### Protocol
+
+```
+                  expire
+      unclaimed <-------- revoked
+          | ^              ^^
+          | |              || expire
+          | | expire       ||
+pre-claim | |              ||  _
+          | |       revoke || | | transfer
+          v |              || | v
+     pre-claimed -------> claimed
+                  claim    | ^   
+                           | |
+                            -
+                          update
+```
+
+Note that `expire` is not an explicit message that is part
+of the protocol.
+
+
+#### Pre-claim
+
+Send a pre-claim transaction containing a hash commitment.
+
+```
+ ------------ ----
+| hash       | 32 |
+ ------------ ----
+```
+
+
+
+#### Claim
+
+```
+ ------------ ----
+| name       | 63 |
+ ------------ ----
+| fee        | 32 |
+ ------------ ----
+```
```

Comment:
does the claim have an expiration as well? otherwise can't it be claimed forever and not be updated ever, to block the name?

Can't we combine the `claim` and `update` so a first claim would include the initial data and an update for renewal purposes would just claim the same name for the same owner again to push back the expiry

Reply:
The `claim` transaction just finishes the claiming process, i.e. after it is in the chain, the name will be regarded as `claimed`.
So this `claim` transaction itself has no expiration date but the claimed name does.